File: frontend/src/arkham/types/Difficulty.ts
import * as JsonDecoder from "ts.data.json";

export type Difficulty = 'Easy' | 'Standard' | 'Hard' | 'Expert';

export const difficultyDecoder = JsonDecoder.oneOf<Difficulty>([
    JsonDecoder.literal('Easy'),
    JsonDecoder.literal('Standard'),
    JsonDecoder.literal('Hard'),
    JsonDecoder.literal('Expert')
], 'Difficulty');

File: backend/arkham-core/library/Arkham/Difficulty.hs
module Arkham.Difficulty where

import Arkham.Prelude

data Difficulty
  = Easy
  | Standard
  | Hard
  | Expert
  deriving stock (Eq, Show, Generic)
  deriving anyclass (ToJSON, FromJSON, NoThunks, NFData)

fromDifficulty :: a -> a -> a -> a -> Difficulty -> a
fromDifficulty easy standard hard expert = \case
  Easy -> easy
  Standard -> standard
  Hard -> hard
  Expert -> expert
